### PR TITLE
Allow user to set content-type header when using JSON output format families

### PR DESCRIPTION
**Pull request by larryluogit**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow user to set content-type header when using JSON output format families

### Documentation entry for user-facing changes

- [X] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit cee75b237802e72e9d98726e0b9f924273ee2b8c with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65254/cee75b237802e72e9d98726e0b9f924273ee2b8c/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by den-crane** [comment]
Wondering why only JSONEachRow? What if I want to set it for TSV as `text\tsv` ?

**Comment by larryluogit** [comment]
> Wondering why only JSONEachRow? What if I want to set it for TSV as `text\tsv` ?

The reason is that JSON supports multiple content types. Deciding the content type purely based on the output_format_json_array_of_rows setting is not enough. 
The content type of TSV is set to 'text/tab-separated-values; charset=UTF-8' currently, and it is correct.

File: src/Processors/Formats/Impl/JSONEachRowRowOutputFormat.h [hunk 1]
```diff
@@ -27,6 +27,8 @@ class JSONEachRowRowOutputFormat : public RowOutputFormatWithExceptionHandlerAda
     /// Content-Type to set when sending HTTP response.
     String getContentType() const override
     {
+        if (!settings.json.content_type.empty())
+            return settings.json.content_type;
```
Comment (Avogar) [inline src/Processors/Formats/Impl/JSONEachRowRowOutputFormat.h:31]: We have several others `JSON*` formats, like `JSON/JSONColumns/JSONCompactColumns/JSONCompactEachRow`. Maybe better to do it for all of them and not only for `JSONEachRow`?

Comment (larryluogit) [reply src/Processors/Formats/Impl/JSONEachRowRowOutputFormat.h:31]: All JSON output format are supported now.

File: src/Processors/Formats/Impl/JSONColumnsBlockOutputFormatBase.h [hunk 2]
```diff
@@ -20,6 +20,13 @@ class JSONColumnsBlockOutputFormatBase : public OutputFormatWithUTF8ValidationAd
 
     String getName() const override { return "JSONColumnsBlockOutputFormatBase"; }
 
+    String getContentType() const override
+    {
+        if (!format_settings.json.content_type.empty())
+            return format_settings.json.content_type;
+        return OutputFormatWithUTF8ValidationAdaptor::getContentType();
```
Comment (Avogar) [inline src/Processors/Formats/Impl/JSONColumnsBlockOutputFormatBase.h:27]:
```suggestion
        return "application/json; charset=UTF-8";
```

It was actually a mistake that this class didn't have implementation of `getContentType`

File: tests/queries/0_stateless/02935_http_content_type_setter_for_json.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+# for frmt in JSONStrings JSON JSONEachRow JSONColumns JSONColumnsWithMetadata JSONCompact JSONCompactStrings JSONCompactColumns PrettyJSONEachRow JSONEachRowWithProgress JSONStringsEachRow JSONStringsEachRowWithProgress JSONCompactEachRow JSONCompactEachRowWithNames JSONCompactEachRowWithNamesAndTypes JSONCompactStringsEachRow JSONCompactStringsEachRowWithNames JSONCompactStringsEachRowWithNamesAndTypes JSONObjectEachRow
+for frmt in JSONStrings JSON JSONEachRow JSONColumns JSONCompact JSONObjectEachRow
+do
+  echo $frmt
+  url="${CLICKHOUSE_URL}/&output_format_json_content_type_header=custom-type&query=select+sleepEachRow(0.01)from+numbers(10)+FORMAT+${frmt}"
```
Comment (Avogar) [inline tests/queries/0_stateless/02935_http_content_type_setter_for_json.sh:11]: Why do we need `sleepEachRow` here?

Comment (larryluogit) [reply tests/queries/0_stateless/02935_http_content_type_setter_for_json.sh:11]: no need. It's cloned from 02935_http_content_type_with_progress.

Comment (Avogar) [reply tests/queries/0_stateless/02935_http_content_type_setter_for_json.sh:11]: Let's remove, it slows down the test

**Comment by Avogar** [comment]
@jsc0218 sorry for taking over the PR, as I already reviewed it, I will assign it to me

**Review by Avogar — APPROVED** [review]

File: src/Core/SettingsChangesHistory.h [hunk 4]
```diff
@@ -115,6 +115,7 @@ static const std::map<ClickHouseVersion, SettingsChangesHistory::SettingsChanges
               {"output_format_csv_serialize_tuple_into_separate_columns", true, true, "A new way of how interpret tuples in CSV format was added."},
               {"input_format_csv_deserialize_separate_columns_into_tuple", true, true, "A new way of how interpret tuples in CSV format was added."},
               {"input_format_csv_try_infer_strings_from_quoted_tuples", true, true, "A new way of how interpret tuples in CSV format was added."},
+              {"output_format_json_content_type_header", "", "", "Add settings to allow user to set Content-Type header when using `JSONEachRow` output format in URL function"},
```
Comment (Avogar) [inline src/Core/SettingsChangesHistory.h:118]: We already created a branch for 24.6 release, let's move it to 24.7 and merge

**Comment by Avogar** [comment]
@larryluogit can you please update `SettingsChangesHistory` file? I don't have permissions to commit in your fork

**Comment by alexey-milovidov** [comment]
Closing in favor of #59620.

**Comment by larryluogit** [comment]
@alexey-milovidov 
The problem I encountered was that when using a materialized view with the URL engine to post JSON data to an external service, ClickHouse sets the Content-Type header using this logic:

`return settings.json.array_of_rows ? "application/json; charset=UTF-8" : "application/x-ndjson; charset=UTF-8";`

The Content-Type is incorrect when a block of rows needs to be posted as a single JSON object rather than an array. I don't think this can be handled by #59620
The fact that a user can provide the content type for JSON format through a query with this change is a side effect, not the primary objective.
Any suggestions?

**Comment by larryluogit** [comment]
@Avogar can this pr be merged?

**Comment by alexey-milovidov** [comment]
We should analyze the orthogonality with https://github.com/ClickHouse/ClickHouse/issues/59620. It is non obvious.

**Comment by alexey-milovidov** [comment]
This is strange. We already allow any HTTP headers when using the URL engine, with the `headers(name = value)` argument.

**Comment by alexey-milovidov** [comment]
```
$ ch
:) INSERT INTO FUNCTION url('http://localhost:5001/', JSONEachRow, 'x UInt8', headers('Content-Type' = 'upyachka')) SELECT 1
```

```
$ nc -l localhost 5001
POST / HTTP/1.1
Host: localhost
Transfer-Encoding: chunked
Content-Type: application/x-ndjson; charset=UTF-8
Content-Type: upyachka
Connection: Keep-Alive
Keep-Alive: timeout=30, max=1000

8
{"x":1}

0
```

Here you can see that the Content-Type header is duplicated.
This is what you need to fix - the user-provided header should replace the default one.

Please open another pull request.

**Pull request closed without merging**